### DISCUSSION TITLE
The Modbus response is incorrect

### DISCUSSION THREAD

**Discussion by mattia424** [eModbus/eModbus#285] [Q&A] [answered]
I created a test environment consisting of:
FT232 converter - MAX485  <=====> MAX3485 - ESP32-S3 (schematic below)
I use ModbusPoll to send data via the converter connected directly to the PC.
I loaded the example https://github.com/eModbus/eModbus/tree/master/examples/RTUserver modifying it to accept the F05 Read COIL

I verified the argument https://github.com/eModbus/eModbus/discussions/112
and I noticed that slave RX is always low and goes high when RE-DE is high just like https://github.com/eModbus/eModbus/discussions/112#discussioncomment-1539397 says here
I managed to "solve" by reversing the pull-up and pull-down on line A and B (R10 to GND and R9 to VCC) is it correct that you should do so? if you why?

![Schema](https://user-images.githubusercontent.com/99872375/230647283-02c163cd-cc62-45e7-870b-2035522f9c6b.png)

After reversing the resistors I no longer find the error  ` [1;33m[E] 1016307| ModbusServerRTU.cpp  [ 252] serves: RTU receive: E5 - Packet length error ` that I received before entering the callback cable

Now I should fix the problem in the answer, from ModbusPull when I send the Write Single Coil I always find the Result Write error (I think it is because it is not found an answer or is not correct). ESP32 side from the serial I see it performs the callback function properly 

![MODBUSPULL_CONFIG](https://user-images.githubusercontent.com/99872375/230647355-fa510032-a366-4bd6-9553-c45353422cd8.jpg)

![WriteCoil](https://user-images.githubusercontent.com/99872375/230647361-d2ceb7de-ac0e-4d0d-b061-2a4a3d195110.jpg)

```
ModbusMessage FC05(ModbusMessage request) {
  Serial.println("Start F05 function");
  uint16_t address;

  ModbusMessage response;
  request.get(2, address);
  Serial.print("Address: "); Serial.println(address);

  Serial.print("request.getServerID(): ");
  Serial.println(request.getServerID());

  Serial.print("request.getFunctionCode(): ");
  Serial.println(request.getFunctionCode());

  response.add(request.getServerID(), request.getFunctionCode(), ILLEGAL_DATA_ADDRESS);
  
  return response;
} 
```

for now I have removed all controls and I always reply with ILLEGAL_DATA_ADDRESS but this answer does not return to me in ModbusPull?
Where is that wrong? Need more code?

Sorry about the language, but I don’t really know English.

Thanks!!

**Comment by Miq1** [comment]
In your schematics, the pull-up and pull-downs are connected inversely to what is common. The wires are named "A+" and "B-" on purpose to indicate A+ should be pulled up, while B- needs to be pulled down.

This must be done **on one end of the bus only**! You seem to be doing that on both ends, and worse, with different voltages. That most probably will not work.

Please try to fix that first before we have a chance to see if something else may be wrong.

**Comment by mattia424** [comment]
You're right, it's wrong in the diagram.
I checked, the converter side is right with A+ and B- (I use the commercial module TTL-->RS485) while for the MAX3485 side ESP32 at the beginning it was connected like the diagram and so I'll explain why it works only by inverting them! 
Now it works leaving only the resistors present in the RS485 module!

I read in the discussion of https://github.com/eModbus/eModbus/discussions/112 that only one pair of resistors are needed in the bus, why this? if I have to make a slave module, do I have to insert them so that the other modules on the market are compatible? is there a rule of who should put these resistors in the bus? is he directly the installer of the master-slave system?

Do you have any suggestions regarding the incorrect answer?

**Reply by Miq1** [reply]
The bus normally is driven from one end. Only there the pullup/pulldown is required. The devices connected to the bus will need the 120 Ohms termination resistor only.

I must confess I did not understand what you were trying to tell about the wrong responses. May I suggest you write down your question in your language and I will try to use DeepL to translate it?

**Comment by mattia424** [comment, marked as answer]
Hi, thank you for introducing me to this translator which works very well!
I have checked better the problem of the response not reaching my ESP32.... I started from the TX pin of ESP32 until I got to the FT232 converter and I found out that it is the latter that does not work corectually. I had read in some forum that in order to make the FT232 work with the MAX485 biosnave to connect the pins RE DO to the BUS2 PIN and evidently this did not work for me... 
Now I have connected a device that acts as a TCP-Modbus Gateway (Adam 5000) and with this I have not encountered any more problems!
P.S. do you know with what oscilloscope/program the outputs that bertmelis put here https://github.com/eModbus/eModbus/discussions/112#discussioncomment-1539397 were generated ?

Thanks a lot for the help and congratulations really for the library and support !

**Reply by Miq1** [reply]
Great that you got it to work!

Bert (and I am, too) is using a Chinese clone of a Salae logic analyzer and Pulseview as software.
